README.md
# A solution for challenge from Purely Functional TV mailing list.

@see https://purelyfunctional.tv/issues/purelyfunctional-tv-newsletter-352-tip-use-the-right-kind-of-comment-for-the-job/

Write a program that outputs all possibilities to put a + or – or
nothing between the numbers 1-9 (in order) such that the result
is 100. For example:

1 + 2 + 34 - 5 + 67 - 8 + 9 = 100

The function should output a list of strings with these formulas.


## About

The solution is by no means optimal. This is just an exercise on TDD and on Clojure development and exploring Clojure.

I've got lots of fun writing these functions and tests for them.

## Profiling

My first solution appeared incredibly slow: 35 seconds to generate and filter all the options.

#### Tufte
I've accepted the challenge and started to profile. Tufte appeared to be a good tool. One nuance, it wasn't work properly when I put a threading macros inside `profile`.

My initial guess was that the work with collections is suboptimal. Profiling showed that the slowest function is `reduce+-`, which appends signs to the digits and then sum them up.

The further investigation revealed `(apply +)` as the slowest construction. Then it was obvious that either `apply` or `+` is slow. But that is not the case! This confusion was made by laziness of the collections with "resolved" numbers. The very "resolving" numbers was really slow.

#### Try ClojureScript

I realized it only later. But at that moment I was thinking of slow integer math. That was ridiculous and I decided to try run the same code with ClojureScript. And when I pasted it to the REPL `eval` wasn't functioning. So I replaced `(comp eval seq)` with `(fn [[f x]] (f x))`, and it running half a second. A lot faster than half a minute.

#### Narrow Down

At that point I wasn't sure what is the reason: ClojureScript or replaced `(comp eval seq)`. Then I measured the difference between two options:

```
(perf (comp eval seq))
; Elapsed time: 1530.837363 msecs

(perf (fn [[op digit]] (op digit)))
; Elapsed time: 0.860691 msecs
```

More than a thousand time of difference. Either `eval` or `seq` is really slow.

#### The Answer
And what is the root of the slowness?

```
pId            nCalls      50% ≤       Mean      Clock  Tota
defn_-eval      1,000     1.60ms     1.63ms     1.63s     99%
defn_-seq       1,000     1.20μs     1.64μs     1.64ms     0%
```

## Usage

Start tests every time code or tests changed:
``` sh
lein auto test
```

## License

Copyright © 2019 shofel

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-11-19
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-11-19
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clj_digitsto100/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj_digitsto100/compare/0.1.0...0.1.1


test/clj_digitsto100/core_test.clj
(ns clj-digitsto100.core-test
  (:require [clojure.test :refer :all]
            [clj-digitsto100.core :refer :all]))


(deftest decode|-test
  (testing "No |"
    (is (= (decode| []) []))
    (is (= (decode| [3]) [3])))
  (testing "Applies | correctly"
    (is (= (decode| [1 | 1]) [11]))
    (is (= (decode| [1 | 2 | 3]) [123])))
  (testing "Preserves + and -"
    (is (= (decode| [1 + 1]) [1 + 1]))
    (is (= (decode| [1 - 1]) [1 - 1])))
  (testing "Combinations of | + -"
    (is (= (decode| [1 | 2 - 3]) [12 - 3]))
    (is (= (decode| [1 | 2 + 3]) [12 + 3])))
  (testing "When no | at the first place"
    (is (= (decode| [1 + 2 | 3]) [1 + 23]))))


(deftest sum'-test
  (testing "all?"
    (is (= (sum' [+ 1]) 1))
    (is (= (sum' [- 1]) -1))
    (is (= (sum' [- 1 - 2]) -3))))


(deftest reduce+--test
  (testing "Just a value"
    (is (= (reduce+- []) 0))
    (is (= (reduce+- [1]) 1))
    (is (= (reduce+- [1 - 2]) -1))))


(deftest decode-test
  (testing "Let's start with |"
    (is (= (decode [1 | 2]) 12))
    (is (= (decode [1 | 2 | 3]) 123)))
  (testing "+ and -"
    (is (= (decode [1 + 2 - 3]) 0)))
  (testing "The right precedence | vs +"
    (is (= (decode [1 + 2 | 3]) 24))
    (is (= (decode [1 + 2 | 3 - 4]) 20))))


(deftest |-test
  (testing "Only trivial cases. No leading zero; only single digits"
    (is (= (| 1 2) 12))
    (is (= (| 3 5) 35))
    (is (= (| 6 0) 60))))

(deftest hundred?-test
  (testing "A solution from the text of the task"
    (is (= (hundred? [1 + 2 + 3 | 4 - 5 + 6 | 7 - 8 + 9]) true)))
  (testing "Obviously wrong: sum all the digits"
    (is (= (hundred? [1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9]) false))))



(deftest append-tails-test
  (testing "??"
    (is (= (append-tails [1 2 3] [7 8])
           [[1 2 3 7] [1 2 3 8]]))))

(deftest combinations-test
  (testing "The empty combination of zero-length"
    (is (= (combinations 0 []) [[]])))
  (testing "Length 1 is each letter"
    (is (= (combinations 1 [6 7]) [[6] [7]])))
  (testing "Real combos"
    (is (= (set (combinations 2 [1 2]))
           #{[1 1] [1 2] [2 1] [2 2]})))
  (testing "Length of the list"
    (is (= (count (combinations 3 [1 2 3])) 27))
    (is (= (count (combinations 4 [1 2])) 16))))


(deftest opt->expression-test
  (testing "Augment ops with digits"
    (is (= (ops->expression [+ - |])
           [1 + 2 - 3 | 4]))))

;;; to string

(deftest render-one-expression-test
  (testing "render expression"
    (is (= (->
             [- | +]
             ops->expression
             render-expression)
           "1 - 23 + 4"))))

(deftest render-single-solution-test
  (testing "render one solution"
    (is (= (->
             [- | +]
             ops->expression
             render-single-solution)
           "1 - 23 + 4 = 100"))))

(deftest expr->chars-test
  (testing "expr to chars"
    (is (= (expr->chars [1 + 2]) [1 " + " 2]))))


src/clj_digitsto100/core.clj
;;;
;;; @see https://purelyfunctional.tv/issues/purelyfunctional-tv-newsletter-352-tip-use-the-right-kind-of-comment-for-the-job/
;;;
;;; Write a program that outputs all possibilities to put a + or – or
;;; nothing between the numbers 1-9 (in order) such that the result
;;; is 100. For example:
;;;
;;; 1 + 2 + 34 - 5 + 67 - 8 + 9 = 100
;;;
;;; The function should output a list of strings with these formulas.
;;;

(ns clj-digitsto100.core
  (:require [clojure.string :as str]
            [taoensso.tufte
             :as tufte
             :refer (defnp p profile)]))

;;; As task mentions, the output should be a list of strings.
;;; Let's make it as the last step.
;;; Then what's the first?

;; Basically, we should make a couple of things:
;;   - the first is decision taking: choose one of three options
;;   - provide inputs for those decisions
;;   - accumulate taken decisions

;; Ok, then. The input is digits from 1 to 9.
;; What is the output?
;;   - a list of all decisions taken?
;;   - a list of digits and signs between them?
;;
;; The first option will require merging back with digits.
;; The second option contain all the info. But `|` changes the number
;; of elements.
;;
;; To not be affected by this, let's use a sequence with all
;; the digits and ops between them.

(defn |
  "An operator which represents decision to glue two digits.
  @example (= (| 1 2) 12)"
  [x y]
  (+ (* 10 x) y))

;;;
;;; Decode |
;;;

(declare decode|')

(defn decode|
  "Apply `decode|'` to a list"
  [x]
  (apply decode|' x))

(defnp decode|'
  "Make expression with all the `|` applied"
  ([] [])
  ([x] [x])
  ([x op y & tail]
   (condp = op
     | (decode| (cons (| x y) tail))
     (concat [x op] (decode| (cons y tail))))))

;;;
;;; Decode + -
;;;

(defn sum'
  "Given a list like [- 2 + 3 ...],
  make a list of numbers with + or - sign."
  [xs]
  (->> xs
    (partition-all 2)
    (map (fn [[op digit]] (op digit)))
    (apply +)))

(comment
  ;; perf of sum(map (comp eval seq))ming [- 2 + 3 ...]

  (defn sample
    "Pairs of a sign and a number."
    [n]
    (repeat n [+ n]))

  (defn perf
    [f]
    (time (dotimes [_ 100]
            (->> (sample 10)
                 (map f)
                 (apply +)))) )

  ; Elapsed time: 1530.837363 msecs
  (perf (comp eval seq))

  ; Elapsed time: 0.860691 msecs
  (perf (fn [[op digit]] (op digit)))

  (defnp -eval [& args] (apply eval args))
  (defnp -seq [& args] (apply seq args))

  ;; What is slow:`comp` `eval` or `seq` ??
  ;;
  ;; pId            nCalls      50% ≤       Mean      Clock  Total
  ;; defn_-eval      1,000     1.60ms     1.63ms     1.63s     99%
  ;; defn_-seq       1,000     1.20μs     1.64μs     1.64ms     0%
  (profile
    {}
    (dotimes [_ 100]
      (let [xs (sample 10)
            numbers (map
                      (comp -eval -seq)
                      xs)
            sum (apply + numbers)]
        sum)))
  )

(declare reduce+-')

;; TODO find the better way
;;      - a function in core?
;;      - pattern matching syntax?
(defn reduce+-
  "Adapter for reduce+-' for easier pattern matching"
  [x]
  (apply reduce+-' x))

(defnp plus
  [& xs]
  (apply + xs))

(defnp reduce+-'
  "Apply + and - in a given seq."
  ([] 0)
  ([x] x)
  ([x & xs]
   (p :plus-body (+ x (sum' xs)))))

;;;
;;; Combine `decode|` and `reduce+-`.
;;; It's the final decode.
;;;

(defn decode
  "Which number is represented by the sequence?"
  [xs]
  (-> xs
      decode|
      reduce+-))

;; It'd be easier to have two different lists:
;; - one with digits from 1 to 9
;; - one with taken decisions
;; Then zip them, and reduce to a number.
(defn hundred?
  "Check if all taken decisions lead to the sum of 100"
  [expression]
  (-> expression
      decode
      (= 100)))



;;; At this point we are ready to get the right solutions from a list
;;; of solutions. Let's generate all solutions and take correct ones.

(defn append-tails
  [head tails]
  (let [heads (repeat (count tails) head)]
    (map conj heads tails)))

(defn combinations
  "Generate all combinations for a given alphabet of a given length"
  ([length alphabet]
   (condp = length
     (combinations [[]] length alphabet)))
  ([heads length alphabet]
   (if (zero? length) heads
     (combinations
       (mapcat
         #(append-tails % alphabet)
         heads)
       (dec length)
       alphabet))))


;;; TODO seq of ops to expression
;;; TODO generate and filter

(defn ops->expression
  [xs]
  (concat [1]
          (mapcat vector xs (range 2 10))))

(defn expr->chars
  [xs]
  (let [opchars {- " - "
                 + " + "
                 | "" }]
    (map
      #(get opchars % %)
      xs)))

(defn render-expression
  [expr]
  (-> expr
      expr->chars
      (str/join)))

(defn render-single-solution
  [s]
  (-> s
      render-expression
      (str " = 100")))

;;; TODO make expression ops always looking well

(defn -main
  []
  (->>
    (combinations 8 [+ | -])
    (map ops->expression)
    (filter hundred?)
    (map render-single-solution)
    (map println)))

#_(-main)

;;; Profile
(tufte/add-basic-println-handler!
{:format-pstats-opts {:columns [:n-calls :p50 :mean :clock :total]
                      :format-id-fn name}})

;;; perf

(comment
  (defn -combos
    []
    (combinations 8 [+ | -]))

  ;; Combinations is fast, lets measure the rest.

  ; 0s: -combos
  (time (->> (-combos)
             doall
             (take 0)))

  (def combos (-combos))

  ;; 0s: ops->expression
  (time (->> combos
             (map ops->expression)
             doall
             (take 0)))

  ;; ?s: sum numbers
  (time (dotimes [n 10e3]
          (sum' (interleave
                  (repeat n :+)
                  (repeat n n)))))
  )

#_
(do
  (defn expressions
    []
    (map ops->expression
         (combinations 8 [+ | -])))

  ;; 35s: filter hundred?
  (profile
    {}
    (take 0 (doall
              (filter hundred? (expressions))))))

;; Answer
#_((1 + 2 + 3 - 4 + 5 + 6 + 7 | 8 + 9)
   (1 + 2 + 3 | 4 - 5 + 6 | 7 - 8 + 9)
   (1 + 2 | 3 - 4 + 5 + 6 + 7 | 8 - 9)
   (1 + 2 | 3 - 4 + 5 | 6 + 7 + 8 + 9)
   (1 | 2 + 3 + 4 + 5 - 6 - 7 + 8 | 9)
   (1 | 2 + 3 - 4 + 5 + 6 | 7 + 8 + 9)
   (1 | 2 - 3 - 4 + 5 - 6 + 7 + 8 | 9)
   (1 | 2 | 3 + 4 - 5 + 6 | 7 - 8 | 9)
   (1 | 2 | 3 + 4 | 5 - 6 | 7 + 8 - 9)
   (1 | 2 | 3 - 4 - 5 - 6 - 7 + 8 - 9)
   (1 | 2 | 3 - 4 | 5 - 6 | 7 + 8 | 9))

;; Formatted answer
#_(list
    "1 + 2 + 3 - 4 + 5 + 6 + 78 + 9 = 100"
    "1 + 2 + 34 - 5 + 67 - 8 + 9 = 100"
    "1 + 23 - 4 + 5 + 6 + 78 - 9 = 100"
    "1 + 23 - 4 + 56 + 7 + 8 + 9 = 100"
    "12 + 3 + 4 + 5 - 6 - 7 + 89 = 100"
    "12 + 3 - 4 + 5 + 67 + 8 + 9 = 100"
    "12 - 3 - 4 + 5 - 6 + 7 + 89 = 100"
    "123 + 4 - 5 + 67 - 89 = 100"
    "123 + 45 - 67 + 8 - 9 = 100"
    "123 - 4 - 5 - 6 - 7 + 8 - 9 = 100"
    "123 - 45 - 67 + 89 = 100")
